File: packages/katala/core/SynergyScorer.ts
export interface PhoenixScores {
  favorite_score?: number;
  reply_score?: number;
  retweet_score?: number;
  photo_expand_score?: number;
  click_score?: number;
  profile_click_score?: number;
  vqv_score?: number;
  share_score?: number;
  share_via_dm_score?: number;
  share_via_copy_link_score?: number;
  dwell_score?: number;
  quote_score?: number;
  quoted_click_score?: number;
  dwell_time?: number;
  follow_author_score?: number;
  not_interested_score?: number;
  block_author_score?: number;
  mute_author_score?: number;
  report_score?: number;
}

export const SCORING_WEIGHTS = {
  FAVORITE_WEIGHT: 0.5,
  REPLY_WEIGHT: 1.0,
  RETWEET_WEIGHT: 0.8,
  PHOTO_EXPAND_WEIGHT: 0.2,
  CLICK_WEIGHT: 0.3,
  PROFILE_CLICK_WEIGHT: 0.4,
  VQV_WEIGHT: 0.6,
  SHARE_WEIGHT: 0.7,
  SHARE_VIA_DM_WEIGHT: 1.2,
  SHARE_VIA_COPY_LINK_WEIGHT: 0.9,
  DWELL_WEIGHT: 0.1,
  QUOTE_WEIGHT: 1.1,
  QUOTED_CLICK_WEIGHT: 0.5,
  CONT_DWELL_TIME_WEIGHT: 0.01,
  FOLLOW_AUTHOR_WEIGHT: 2.0,
  NOT_INTERESTED_WEIGHT: -1.0,
  BLOCK_AUTHOR_WEIGHT: -5.0,
  MUTE_AUTHOR_WEIGHT: -3.0,
  REPORT_WEIGHT: -10.0,
};

export class SynergyScorer {
  private weights: typeof SCORING_WEIGHTS;

  constructor(customWeights?: Partial<typeof SCORING_WEIGHTS>) {
    this.weights = { ...SCORING_WEIGHTS, ...customWeights };
  }

  private applyWeight(score: number | undefined, weight: number): number {
    return (score ?? 0) * weight;
  }

  public computeWeightedScore(scores: PhoenixScores): number {
    let combinedScore = 0;

    combinedScore += this.applyWeight(scores.favorite_score, this.weights.FAVORITE_WEIGHT);
    combinedScore += this.applyWeight(scores.reply_score, this.weights.REPLY_WEIGHT);
    combinedScore += this.applyWeight(scores.retweet_score, this.weights.RETWEET_WEIGHT);
    combinedScore += this.applyWeight(scores.photo_expand_score, this.weights.PHOTO_EXPAND_WEIGHT);
    combinedScore += this.applyWeight(scores.click_score, this.weights.CLICK_WEIGHT);
    combinedScore += this.applyWeight(
      scores.profile_click_score,
      this.weights.PROFILE_CLICK_WEIGHT,
    );
    combinedScore += this.applyWeight(scores.vqv_score, this.weights.VQV_WEIGHT);
    combinedScore += this.applyWeight(scores.share_score, this.weights.SHARE_WEIGHT);
    combinedScore += this.applyWeight(scores.share_via_dm_score, this.weights.SHARE_VIA_DM_WEIGHT);
    combinedScore += this.applyWeight(
      scores.share_via_copy_link_score,
      this.weights.SHARE_VIA_COPY_LINK_WEIGHT,
    );
    combinedScore += this.applyWeight(scores.dwell_score, this.weights.DWELL_WEIGHT);
    combinedScore += this.applyWeight(scores.quote_score, this.weights.QUOTE_WEIGHT);
    combinedScore += this.applyWeight(scores.quoted_click_score, this.weights.QUOTED_CLICK_WEIGHT);
    combinedScore += this.applyWeight(scores.dwell_time, this.weights.CONT_DWELL_TIME_WEIGHT);
    combinedScore += this.applyWeight(
      scores.follow_author_score,
      this.weights.FOLLOW_AUTHOR_WEIGHT,
    );
    combinedScore += this.applyWeight(
      scores.not_interested_score,
      this.weights.NOT_INTERESTED_WEIGHT,
    );
    combinedScore += this.applyWeight(scores.block_author_score, this.weights.BLOCK_AUTHOR_WEIGHT);
    combinedScore += this.applyWeight(scores.mute_author_score, this.weights.MUTE_AUTHOR_WEIGHT);
    combinedScore += this.applyWeight(scores.report_score, this.weights.REPORT_WEIGHT);

    return combinedScore;
  }

  /**
   * Computes synergy score between two agents based on their interests.
   * This is the core logic adapted for agent-to-agent synergy.
   */
  public computeSynergy(interestsA: Map<string, number>, interestsB: Map<string, number>): number {
    let synergy = 0;
    const allCategories = new Set([...interestsA.keys(), ...interestsB.keys()]);

    for (const category of allCategories) {
      const weightA = interestsA.get(category) ?? 0;
      const weightB = interestsB.get(category) ?? 0;
      // Simple dot product for synergy
      synergy += weightA * weightB;
    }

    return synergy;
  }
}

File: packages/katala/core/synergy_scorer.rs
use std::collections::HashMap;

pub struct PhoenixScores {
    pub favorite_score: Option<f64>,
    pub reply_score: Option<f64>,
    pub retweet_score: Option<f64>,
    pub photo_expand_score: Option<f64>,
    pub click_score: Option<f64>,
    pub profile_click_score: Option<f64>,
    pub vqv_score: Option<f64>,
    pub share_score: Option<f64>,
    pub share_via_dm_score: Option<f64>,
    pub share_via_copy_link_score: Option<f64>,
    pub dwell_score: Option<f64>,
    pub quote_score: Option<f64>,
    pub quoted_click_score: Option<f64>,
    pub dwell_time: Option<f64>,
    pub follow_author_score: Option<f64>,
    pub not_interested_score: Option<f64>,
    pub block_author_score: Option<f64>,
    pub mute_author_score: Option<f64>,
    pub report_score: Option<f64>,
}

pub struct ScoringWeights {
    pub favorite_weight: f64,
    pub reply_weight: f64,
    pub retweet_weight: f64,
    pub photo_expand_weight: f64,
    pub click_weight: f64,
    pub profile_click_weight: f64,
    pub vqv_weight: f64,
    pub share_weight: f64,
    pub share_via_dm_weight: f64,
    pub share_via_copy_link_weight: f64,
    pub dwell_weight: f64,
    pub quote_weight: f64,
    pub quoted_click_weight: f64,
    pub cont_dwell_time_weight: f64,
    pub follow_author_weight: f64,
    pub not_interested_weight: f64,
    pub block_author_weight: f64,
    pub mute_author_weight: f64,
    pub report_weight: f64,
}

impl Default for ScoringWeights {
    fn default() -> Self {
        Self {
            favorite_weight: 0.5,
            reply_weight: 1.0,
            retweet_weight: 0.8,
            photo_expand_weight: 0.2,
            click_weight: 0.3,
            profile_click_weight: 0.4,
            vqv_weight: 0.6,
            share_weight: 0.7,
            share_via_dm_weight: 1.2,
            share_via_copy_link_weight: 0.9,
            dwell_weight: 0.1,
            quote_weight: 1.1,
            quoted_click_weight: 0.5,
            cont_dwell_time_weight: 0.01,
            follow_author_weight: 2.0,
            not_interested_weight: -1.0,
            block_author_weight: -5.0,
            mute_author_weight: -3.0,
            report_weight: -10.0,
        }
    }
}

pub struct SynergyScorer {
    weights: ScoringWeights,
}

impl SynergyScorer {
    pub fn new(weights: Option<ScoringWeights>) -> Self {
        Self {
            weights: weights.unwrap_or_default(),
        }
    }

    fn apply_weight(&self, score: Option<f64>, weight: f64) -> f64 {
        score.unwrap_or(0.0) * weight
    }

    pub fn compute_weighted_score(&self, scores: &PhoenixScores) -> f64 {
        let mut combined_score = 0.0;

        combined_score += self.apply_weight(scores.favorite_score, self.weights.favorite_weight);
        combined_score += self.apply_weight(scores.reply_score, self.weights.reply_weight);
        combined_score += self.apply_weight(scores.retweet_score, self.weights.retweet_weight);
        combined_score += self.apply_weight(scores.photo_expand_score, self.weights.photo_expand_weight);
        combined_score += self.apply_weight(scores.click_score, self.weights.click_weight);
        combined_score += self.apply_weight(scores.profile_click_score, self.weights.profile_click_weight);
        combined_score += self.apply_weight(scores.vqv_score, self.weights.vqv_weight);
        combined_score += self.apply_weight(scores.share_score, self.weights.share_weight);
        combined_score += self.apply_weight(scores.share_via_dm_score, self.weights.share_via_dm_weight);
        combined_score += self.apply_weight(scores.share_via_copy_link_score, self.weights.share_via_copy_link_weight);
        combined_score += self.apply_weight(scores.dwell_score, self.weights.dwell_weight);
        combined_score += self.apply_weight(scores.quote_score, self.weights.quote_weight);
        combined_score += self.apply_weight(scores.quoted_click_score, self.weights.quoted_click_weight);
        combined_score += self.apply_weight(scores.dwell_time, self.weights.cont_dwell_time_weight);
        combined_score += self.apply_weight(scores.follow_author_score, self.weights.follow_author_weight);
        combined_score += self.apply_weight(scores.not_interested_score, self.weights.not_interested_weight);
        combined_score += self.apply_weight(scores.block_author_score, self.weights.block_author_weight);
        combined_score += self.apply_weight(scores.mute_author_score, self.weights.mute_author_weight);
        combined_score += self.apply_weight(scores.report_score, self.weights.report_weight);

        combined_score
    }

    pub fn compute_synergy(&self, interests_a: &HashMap<String, f64>, interests_b: &HashMap<String, f64>) -> f64 {
        let mut synergy = 0.0;
        for (category, weight_a) in interests_a {
            if let Some(weight_b) = interests_b.get(category) {
                synergy += weight_a * weight_b;
            }
        }
        synergy
    }
}
